File: C++/Player.cpp
#include "Player.h"
#include <iostream>

Player::Player() : Entity() {
	Health = 50;
	Defense = 5;
	Level = 1;
	XP = 0;
	Name = "Player";
	Symbol = '@';
	Attack = 5;
	x = 10;
	y = 10;
}

bool Player::isAlive() {
	return Health > 0;
}

void Player::setInput(char newInput) {
	input = newInput;
}

void Player::onLevelUp() {
	Attack += 1;
	Health += 10;
}

int* Player::getLevelPtr() {
	return &Level; // Возвращаем указатель на уровень
}

int& Player::getHealthRef() {
	return Health; // Возвращаем ссылку на здоровье
}

void Player::gainXP(int amount) {
	XP += amount;
	if (XP % 100 > Level) {
		Level += 1;
		onLevelUp();
	}
}

std::ostream& operator<<(std::ostream& os, const Player& player) {
	os << "Player XP: " << player.XP 
		<< "\nPlayer level: " << player.Level << "\n";
	return os;
}

File: Java/Player.java
package Java;

import java.util.ArrayList;

public class Player extends Entity implements Killable {
    private int xp;
    private char input;

    public Player() {
        super("Player", 50, 5, 5, 1, '@');
        this.xp = 0;
        this.x = 10;
        this.y = 10;
    }

    @Override
    public int getHealth() {
        return health;
    }

    @Override
    public boolean isAlive() {
        return health > 0;
    }

    public void setInput(char newInput) {
        input = newInput;
    }

    @Override
    protected void onLevelUp() {
        this.defense += 1;
        this.attack += 1;
        this.health += 10;
    }

    public void gainXP(int amount) {
        xp += amount;
        if (xp >= level * 100) {
            level += 1;
            onLevelUp();
        }
    }

    @Override
    public String toString() {
        return "Player XP: " + xp + "\n" +
               "Player level: " + level + "\n";
    }
}
